core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/RootVariableSource.java

```diff
@@ -0,0 +1,246 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+import ai.timefold.solver.core.api.domain.variable.ShadowVariable;
+import ai.timefold.solver.core.config.util.ConfigUtils;
+import ai.timefold.solver.core.impl.domain.common.ReflectionHelper;
+import ai.timefold.solver.core.impl.domain.common.accessor.MemberAccessor;
+import ai.timefold.solver.core.impl.domain.common.accessor.MemberAccessorFactory;
+import ai.timefold.solver.core.impl.domain.entity.descriptor.EntityDescriptor;
+import ai.timefold.solver.core.impl.domain.policy.DescriptorPolicy;
+import ai.timefold.solver.core.preview.api.domain.variable.declarative.ShadowVariableUpdater;
+
+public record RootVariableSource<Entity_, Value_>(
+        Class<? extends Entity_> rootEntity,
+        BiConsumer<? extends Entity_, Consumer<Value_>> valueEntityFunction,
+        List<VariableSourceReference> variableSourceReferences) {
+
+    private static final String COLLECTION_REFERENCE_SUFFIX = "[]";
+    private record VariablePath(Class<?> variableEntityClass,
+            String variableName,
+            List<MemberAccessor> memberAccessorsBeforeEntity,
+            List<MemberAccessor> memberAccessorsAfterEntity) {
+        public BiConsumer<Object, Consumer<Object>> getValueVisitorFromVariableEntity() {
+            return (entity, consumer) -> {
+                var currentEntity = entity;
+                for (var member : memberAccessorsAfterEntity) {
+                    currentEntity = member.executeGetter(currentEntity);
+                    if (currentEntity == null) {
+                        return;
+                    }
+                }
+                consumer.accept(currentEntity);
+            };
+        }
+    }
+
+    public static <Entity_, Value_> RootVariableSource<Entity_, Value_> from(
+            Class<? extends Entity_> rootEntityClass,
+            String targetVariableName,
+            String variablePath,
+            MemberAccessorFactory memberAccessorFactory,
+            DescriptorPolicy descriptorPolicy) {
+        var pathParts = variablePath.split("\\.");
+        var variablePaths = new ArrayList<VariablePath>();
+        List<MemberAccessor> chainToVariable = new ArrayList<>();
+        List<MemberAccessor> listMemberAccessors = new ArrayList<>();
+        var hasListMemberAccessor = false;
+        List<List<MemberAccessor>> chainAfterVariable = new ArrayList<>();
+        boolean afterVariable = false;
+        Class<?> currentEntity = rootEntityClass;
+
+        for (var pathPart : pathParts) {
+            if (pathPart.endsWith(COLLECTION_REFERENCE_SUFFIX)) {
+                if (afterVariable) {
+                    throw new IllegalArgumentException("Cannot reference a collection on a variable.");
+                }
+                if (hasListMemberAccessor) {
+                    throw new IllegalArgumentException("Cannot reference a collection on a collection.");
```

Comment:
Deserves a more descriptive message.
Where did the user make a mistake? Maybe there is a suggestion for fixing it?

Reply:
Also applies to the previous exception immediately above.